torch/utils/checkpoint.py

```diff
@@ -1098,6 +1107,166 @@ def unpack_hook_with_error_cb(holder):
             super().__init__(pack_hook, unpack_hook)
 
 
+def _is_compiling(func, args, kwargs):
+    # Check if we are under AOTAutograd tracing
+    for arg in args:
+        if isinstance(arg, torch.Tensor):
+            if torch._is_functional_tensor(arg):
+                arg = torch._from_functional_tensor(arg)
+            if isinstance(arg, torch._subclasses.FakeTensor):
+                return True
+    return False
+
+
+def _detach(x):
+    if isinstance(x, torch.Tensor):
+        return x.detach()
+    return x
+
+
+uid = count(1)
+
+
+# NOTE: torch.utils.checkpoint internal logic will call these two functions unknown number of times
+# (i.e. there could be _CachedTorchDispatchMode calls that doesn't map to a _CachingTorchDispatchMode call),
+# so we ignore these ops and just always recompute them.
+_ignored_ops = {
+    torch.ops.prim.device.default,
+    torch.ops.aten.detach.default,
+}
+
+
+class _CachingTorchDispatchMode(TorchDispatchMode):
+    r"""
+    A :class:`TorchDispatchMode` to implement selective activation checkpointing
+    that's compatible with torch.compile. Used together with _CachedTorchDispatchMode.
+    """
+    def __init__(self, policy_fn, storage):
+        self.policy_fn = policy_fn
+        self.storage = storage
+
+    def push_into_storage(self, out, func, args, kwargs):
+        out_detached = tree_map(_detach, out)
+        self.storage[func].append(out_detached)
+
+    def _handle_compile_in_forward_ctx(self, should_not_recompute, func, args, kwargs):
+        assert not func._schema.is_mutable, (
+            "In-place ops are not supported in selective checkpointing region under torch.compile. "
+            + f"Found in-place ops: {func}."
+        )
+        assert torch.Tag.nondeterministic_seeded not in func.tags, (
+            "Random ops are not supported in selective checkpointing region under torch.compile. "
+            + f"Found random ops: {func}."
+        )
```

Comment:
Oh is that always the case? Can we never have random ops in selective checkpointing? Why is that the case?

The whole point of Duplicate Late is to handle rng ops. If no rng ops, Duplicate early + disabling cse might be good enough. So, this complexity of TorchDispatchMode and still using AOT Autograd for tracing seems unnecessary. I am still behind the idea of unifying but currently the stack seems little jarring.

Reply:
Removed restriction on random ops